File: ios/RSSignatureCaptureTests/RSSignatureViewTests.m
#import <XCTest/XCTest.h>
#import "RSSignatureView.h"
#import "PPSSignatureView.h"

@interface RSSignatureViewTests : XCTestCase
@property (nonatomic, strong) RSSignatureView *rsView;
@end

@implementation RSSignatureViewTests

- (void)setUp {
    [super setUp];
    self.rsView = [[RSSignatureView alloc] init];
}

- (void)tearDown {
    self.rsView = nil;
    [super tearDown];
}

- (void)testInit_returnsNonNil {
    XCTAssertNotNil(self.rsView);
}

- (void)testDefaults_showBorderYES {
    // Access via KVC since _showBorder is an ivar
    NSNumber *showBorder = [self.rsView valueForKey:@"showBorder"];
    XCTAssertTrue(showBorder.boolValue, @"showBorder should default to YES");
}

- (void)testDefaults_showNativeButtonsYES {
    NSNumber *showNativeButtons = [self.rsView valueForKey:@"showNativeButtons"];
    XCTAssertTrue(showNativeButtons.boolValue, @"showNativeButtons should default to YES");
}

- (void)testDefaults_showTitleLabelYES {
    NSNumber *showTitleLabel = [self.rsView valueForKey:@"showTitleLabel"];
    XCTAssertTrue(showTitleLabel.boolValue, @"showTitleLabel should default to YES");
}

- (void)testSetRotateClockwise_storesValue {
    [self.rsView setRotateClockwise:YES];
    NSNumber *val = [self.rsView valueForKey:@"rotateClockwise"];
    XCTAssertTrue(val.boolValue, @"rotateClockwise should be YES after setting");
}

- (void)testSetSquare_storesValue {
    [self.rsView setSquare:YES];
    NSNumber *val = [self.rsView valueForKey:@"square"];
    XCTAssertTrue(val.boolValue, @"square should be YES after setting");
}

- (void)testErase_callsSignErase {
    // Give the view a frame and trigger layout so `sign` gets created
    self.rsView.frame = CGRectMake(0, 0, 320, 480);
    [self.rsView layoutSubviews];

    XCTAssertNotNil(self.rsView.sign, @"sign should be created after layout");

    // Simulate a signature via KVC on the inner PPSSignatureView
    [self.rsView.sign setValue:@YES forKey:@"hasSignature"];
    XCTAssertTrue(self.rsView.sign.hasSignature);

    [self.rsView erase];
    XCTAssertFalse(self.rsView.sign.hasSignature,
                   @"erase should clear the inner signature view");
}

- (void)testFireDragEvent_callsOnDragEventBlock {
    __block BOOL called = NO;
    __block NSDictionary *receivedBody = nil;

    self.rsView.onDragEvent = ^(NSDictionary *body) {
        called = YES;
        receivedBody = body;
    };

    [self.rsView fireDragEvent];

    XCTAssertTrue(called, @"onDragEvent block should be called");
    XCTAssertEqualObjects(receivedBody[@"dragged"], @YES,
                          @"dragged should be YES");
}

@end

File: ios/RSSignatureCaptureTests/RSSignatureViewTests.swift
import XCTest

class RSSignatureViewTests: XCTestCase {

    var rsView: RSSignatureView!

    override func setUp() {
        super.setUp()
        rsView = RSSignatureView()
    }

    override func tearDown() {
        rsView = nil
        super.tearDown()
    }

    func testInit_returnsNonNil() {
        XCTAssertNotNil(rsView)
    }

    func testDefaults_showBorderYES() {
        XCTAssertTrue(rsView.showBorder, "showBorder should default to YES")
    }

    func testDefaults_showNativeButtonsYES() {
        XCTAssertTrue(rsView.showNativeButtons, "showNativeButtons should default to YES")
    }

    func testDefaults_showTitleLabelYES() {
        XCTAssertTrue(rsView.showTitleLabel, "showTitleLabel should default to YES")
    }

    func testSetRotateClockwise_storesValue() {
        rsView.rotateClockwise = true
        XCTAssertTrue(rsView.rotateClockwise, "rotateClockwise should be YES after setting")
    }

    func testSetSquare_storesValue() {
        rsView.square = true
        XCTAssertTrue(rsView.square, "square should be YES after setting")
    }

    func testErase_callsSignErase() {
        rsView.frame = CGRect(x: 0, y: 0, width: 320, height: 480)
        rsView.layoutSubviews()

        XCTAssertNotNil(rsView.sign, "sign should be created after layout")

        rsView.sign?.hasSignature = true
        XCTAssertTrue(rsView.sign?.hasSignature == true)

        rsView.erase()
        XCTAssertFalse(rsView.sign?.hasSignature ?? true,
                       "erase should clear the inner signature view")
    }

    func testSaveImage_firesOnSaveEventWithPayload() {
        rsView.frame = CGRect(x: 0, y: 0, width: 320, height: 480)
        rsView.layoutSubviews()

        XCTAssertNotNil(rsView.sign, "sign should be created after layout")

        rsView.sign?.hasSignature = true
        rsView.sign?.layoutSubviews()

        var receivedBody: [AnyHashable: Any]?
        rsView.onSaveEvent = { body in
            receivedBody = body
        }

        rsView.saveImage()

        XCTAssertNotNil(receivedBody, "onSaveEvent should be called")
        XCTAssertNotNil(receivedBody?["pathName"] as? String,
                        "payload should contain pathName")
        let encoded = receivedBody?["encoded"] as? String
        XCTAssertNotNil(encoded, "payload should contain encoded")
        XCTAssertFalse(encoded?.isEmpty ?? true,
                       "encoded should not be empty")
    }

    func testFireDragEvent_callsOnDragEventBlock() {
        var called = false
        var receivedBody: [AnyHashable: Any]?

        rsView.onDragEvent = { body in
            called = true
            receivedBody = body
        }

        rsView.fireDragEvent()

        XCTAssertTrue(called, "onDragEvent block should be called")
        XCTAssertEqual(receivedBody?["dragged"] as? Bool, true,
                       "dragged should be YES")
    }
}
